### ISSUE TITLE
panic while panicking in spinning up large cluster

### ISSUE THREAD

**Issue by antifuchs** [MaterializeInc/materialize#12396]
### What version of Materialize are you using?

materialize/materialized:unstable-1ca952b14e1a5e09ee98cb0fe17a19ad45feba45

### How did you install Materialize?

Materialize Cloud

### What is the issue?

Launched a `2xlarge` compute cluster in materialize cloud and observed that it's unable to serve `show tables` even when the pods are available very quickly. Looking at the logs, there's a ton of errors of the form,

```
thread 'timely:work-0' panicked at 'MergeQueue poisoned.', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/7ad8962/communication/src/allocator/zero_copy/bytes_exchange.rs:119:52
```

And the occasional panic while panicing. Attaching that panic below.

### Relevant log output

```source
compute-cluster-2-replica-2-7    0:     0x55cdadcf311d - std::backtrace_rs::backtrace::libunwind::trace::hee598835bc88d35b
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
compute-cluster-2-replica-2-7    1:     0x55cdadcf311d - std::backtrace_rs::backtrace::trace_unsynchronized::h9cdc730ba5cf5d72
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
compute-cluster-2-replica-2-7    2:     0x55cdadcf311d - std::sys_common::backtrace::_print_fmt::h75aeaf7ed30e43fa
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:66:5
compute-cluster-2-replica-2-7    3:     0x55cdadcf311d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h606862f787600875
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:45:22
compute-cluster-2-replica-2-7    4:     0x55cdadd1b3bc - core::fmt::write::he803f0f418caf762
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/fmt/mod.rs:1190:17
compute-cluster-2-replica-2-7    5:     0x55cdadceafc8 - std::io::Write::write_fmt::h70bc45872f37e7bb
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/io/mod.rs:1657:15
compute-cluster-2-replica-2-7    6:     0x55cdadcf5477 - std::sys_common::backtrace::_print::h64d038cf8ac3e13e
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:48:5
compute-cluster-2-replica-2-7    7:     0x55cdadcf5477 - std::sys_common::backtrace::print::h359300b4a7fccf65
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:35:9
compute-cluster-2-replica-2-7    8:     0x55cdadcf5477 - std::panicking::default_hook::{{closure}}::hf51be35e2f510149
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:295:22
compute-cluster-2-replica-2-7    9:     0x55cdadcf5140 - std::panicking::default_hook::h03ca0f22e1d2d25e
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:314:9
compute-cluster-2-replica-2-7   10:     0x55cdadcf5cc9 - std::panicking::rust_panic_with_hook::h3b7380e99b825b63
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:698:17
compute-cluster-2-replica-2-7   11:     0x55cdadcf59b7 - std::panicking::begin_panic_handler::{{closure}}::h8e849d0710154ce0
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:588:13
compute-cluster-2-replica-2-7   12:     0x55cdadcf35e4 - std::sys_common::backtrace::__rust_end_short_backtrace::hedcdaddbd4c46cc5
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:138:18
compute-cluster-2-replica-2-7   13:     0x55cdadcf56c9 - rust_begin_unwind
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
compute-cluster-2-replica-2-7   14:     0x55cdadd188f3 - core::panicking::panic_fmt::he1bbc7336d49a357
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
compute-cluster-2-replica-2-7   15:     0x55cdadd18ba3 - core::result::unwrap_failed::h719f957ed74f93fd
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/result.rs:1749:5
compute-cluster-2-replica-2-7   16:     0x55cdacb8051c - core::result::Result<T,E>::expect::h1243a398323c2c01
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/result.rs:1022:23
compute-cluster-2-replica-2-7   17:     0x55cdacb8051c - <timely_communication::initialize::WorkerGuards<T> as core::ops::drop::Drop>::drop::hbfd72948033aff3c
compute-cluster-2-replica-2-7                                at /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/7ad8962/communication/src/initialize.rs:354:13
compute-cluster-2-replica-2-7   18:     0x55cdac044ed0 - core::ptr::drop_in_place<timely_communication::initialize::WorkerGuards<()>>::h3794aeeb7a06ab0f
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ptr/mod.rs:188:1
compute-cluster-2-replica-2-7   19:     0x55cdac044ed0 - core::ptr::drop_in_place<mz_compute::server::Server>::h1b74d85856f2929b
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ptr/mod.rs:188:1
compute-cluster-2-replica-2-7   20:     0x55cdac0663ba - computed::serve::{{closure}}::ha0cb64f7a7f79dd5
compute-cluster-2-replica-2-7                                at /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0d32dbd5c839b263a-1/materialize/tests/src/compute/src/bin/computed.rs:308:1
compute-cluster-2-replica-2-7   21:     0x55cdac0663ba - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h693f21682df1f4fa
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/future/mod.rs:84:19
compute-cluster-2-replica-2-7   22:     0x55cdac0663ba - computed::run::{{closure}}::hf5fafde9181a45d3
compute-cluster-2-replica-2-7                                at /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0d32dbd5c839b263a-1/materialize/tests/src/compute/src/bin/computed.rs:291:40
compute-cluster-2-replica-2-7   23:     0x55cdac0f83b3 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::he86ff6c8e0932983
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/future/mod.rs:84:19
compute-cluster-2-replica-2-7   24:     0x55cdac0f83b3 - computed::main::{{closure}}::hfcfeb25e2fd02383
compute-cluster-2-replica-2-7                                at /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0d32dbd5c839b263a-1/materialize/tests/src/compute/src/bin/computed.rs:131:53
compute-cluster-2-replica-2-7   25:     0x55cdac0f83b3 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h75755cf6db2eedbd
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/future/mod.rs:84:19
compute-cluster-2-replica-2-7   26:     0x55cdac2d481b - tokio::park::thread::CachedParkThread::block_on::{{closure}}::h7f724efd7da3eadb
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/park/thread.rs:263:54
compute-cluster-2-replica-2-7   27:     0x55cdac2d481b - tokio::coop::with_budget::{{closure}}::h9266a6e2df79c66b
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/coop.rs:102:9
compute-cluster-2-replica-2-7   28:     0x55cdac2d481b - std::thread::local::LocalKey<T>::try_with::h59193739e34a7b1a
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/thread/local.rs:413:16
compute-cluster-2-replica-2-7   29:     0x55cdac2d481b - std::thread::local::LocalKey<T>::with::h4bcd5909aba59fd8
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/thread/local.rs:389:9
compute-cluster-2-replica-2-7   30:     0x55cdac2d481b - tokio::coop::with_budget::hef11fc83ab1b4d7a
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/coop.rs:95:5
compute-cluster-2-replica-2-7   31:     0x55cdac2d481b - tokio::coop::budget::hf26a3b9539156b63
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/coop.rs:72:5
compute-cluster-2-replica-2-7   32:     0x55cdac2d481b - tokio::park::thread::CachedParkThread::block_on::h94084fcb3dafe2ba
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/park/thread.rs:263:31
compute-cluster-2-replica-2-7   33:     0x55cdac973c6d - tokio::runtime::enter::Enter::block_on::hf758b3569cd22f78
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/enter.rs:151:13
compute-cluster-2-replica-2-7   34:     0x55cdac973c6d - tokio::runtime::thread_pool::ThreadPool::block_on::h4c8778a744310661
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/thread_pool/mod.rs:73:9
compute-cluster-2-replica-2-7   35:     0x55cdac6a0b06 - tokio::runtime::Runtime::block_on::hd5c6d0964222164f
compute-cluster-2-replica-2-7                                at /cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.17.0/src/runtime/mod.rs:477:43
compute-cluster-2-replica-2-7   36:     0x55cdacb79f1a - computed::main::hb8ed92df8a40dd89
compute-cluster-2-replica-2-7                                at /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0d32dbd5c839b263a-1/materialize/tests/src/compute/src/bin/computed.rs:131:5
compute-cluster-2-replica-2-7   37:     0x55cdac6a08e3 - core::ops::function::FnOnce::call_once::h1ca13cb391a4352c
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:227:5
compute-cluster-2-replica-2-7   38:     0x55cdac6a08e3 - std::sys_common::backtrace::__rust_begin_short_backtrace::he503a4c0fd4395b1
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:122:18
compute-cluster-2-replica-2-7   39:     0x55cdac69efb9 - std::rt::lang_start::{{closure}}::h3cef29eda56e07a1
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/rt.rs:145:18
compute-cluster-2-replica-2-7   40:     0x55cdadcf2801 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hb7014f43484a8b4e
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:259:13
compute-cluster-2-replica-2-7   41:     0x55cdadcf2801 - std::panicking::try::do_call::h7bc9dc436daeb8c7
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:492:40
compute-cluster-2-replica-2-7   42:     0x55cdadcf2801 - std::panicking::try::h653d68a27ff5f175
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:456:19
compute-cluster-2-replica-2-7   43:     0x55cdadcf2801 - std::panic::catch_unwind::h9d739f9f59895e68
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panic.rs:137:14
compute-cluster-2-replica-2-7   44:     0x55cdadcf2801 - std::rt::lang_start_internal::{{closure}}::hf006f2bc7ce22bbe
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/rt.rs:128:48
compute-cluster-2-replica-2-7   45:     0x55cdadcf2801 - std::panicking::try::do_call::hfb39d6df61a2e69f
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:492:40
compute-cluster-2-replica-2-7   46:     0x55cdadcf2801 - std::panicking::try::h13e2d225134958ac
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:456:19
compute-cluster-2-replica-2-7   47:     0x55cdadcf2801 - std::panic::catch_unwind::h3bd49b5a5dfb1a50
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panic.rs:137:14
compute-cluster-2-replica-2-7   48:     0x55cdadcf2801 - std::rt::lang_start_internal::h2ba92edce36c035e
compute-cluster-2-replica-2-7                                at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/rt.rs:128:20
compute-cluster-2-replica-2-7   49:     0x55cdacbd6cb8 - main
compute-cluster-2-replica-2-7   50:     0x7f9fd37edd90 - <unknown>
compute-cluster-2-replica-2-7   51:     0x7f9fd37ede40 - __libc_start_main
compute-cluster-2-replica-2-7   52:     0x55cdabf52029 - <unknown>
compute-cluster-2-replica-2-7                                at /scratch/.build/src/glibc-2.12.1/csu/../sysdeps/x86_64/elf/start.S:103
compute-cluster-2-replica-2-7 thread panicked while panicking. aborting.
```
```

**Comment by antifuchs** [comment]
To repro, I did this:

1. Spin up a materialize platform environment with tag `materialize/materialized:unstable-1ca952b14e1a5e09ee98cb0fe17a19ad45feba45`
2. When it's available, connect to it with psql
3. Run `CREATE CLUSTER foo REPLICA r1 (SIZE '2xlarge'); set cluster = foo ; show tables;`

Observable: the `show tables` doesn't seem to finish; the compute pods are all running; there is a ton of log output from regularly-crashing compute pods about poisoned queues and the above panic.

**Comment by philip-stoev** [comment]
@[antifuchs](https://github.com/antifuchs) all the panics that you see except for one are the result of the "shared fate" of all the computed processes. So the first computed that has panicked caused the rest to panic as well. Would it be possible to find the first panic that happened, or, in other words, the panic that was different from all the others?

**Comment by antifuchs** [comment]
phew. That may be pretty difficult with the k8s log-extraction tools I used to get this. Probably best if we learn how to do this early, so I'll see if I can get a better log stream.

**Comment by philip-stoev** [comment]
@antifuchs yes, I suspect that this operation will need to be performed many times in the future. If the follow-up panics are a problem for log analysis, which I already suspected they will be, do you mind opening a ticket from your name about the fact that the follow-up shutdowns must be cleaner (1 line error message) in order to not to muddle things?

**Comment by antifuchs** [comment]
Update: This was easier than expected. `kubectl logs` has a follow mode (`-f`), and now repro'ing the log becomes all the above steps, but running `kubectl logs --context mzcloud-asf -n environment-792c5fc3-8fb4-48a5-a5f3-0b12f1dc56d4-0 -l compute.materialized.materialize.cloud/type=cluster --all-containers -f` before I CREATE CLUSTER.


Here's the log: [cluster-creation.log](https://github.com/MaterializeInc/materialize/files/8672840/cluster-creation.log)

Unfortunately, it's not easy to get the logs of containers/pods that don't yet exist: kubectl errors when it finds a pod in an unexpected state. Maybe tools exist that do better there; if this log file is not useful I'll see if there's something better out there.

Anyway, it sounds like consensus is that the line where everything goes wrong is: 
```
pod/compute-cluster-2-replica-2-31/default] 2022-05-11T19:20:29.053035688Z thread 'timely:recv-15' panicked at 'assertion failed: read > 0', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/7ad8962/communication/src/allocator/zero_copy/tcp.rs:67:9
```

**Comment by antifuchs** [comment]
So the problem isn't so much the panics, it's retrieving logs from things that happened at arbitrary times in the past (which, if there's lots of output, k8s will discard old log lines), while *also* not allowing us to listen "ahead of time" to logs emitted by pods that match a selector that don't exist yet (and if any pods exist that aren't in Running state, `kubectl logs` will immediately scream and exit)... so yeah, ideally there's a tool that can do this but I haven't found it yet.

**Comment by antifuchs** [comment]
Spun up a cluster with https://github.com/MaterializeInc/materialize/pull/12383 in place, but unfortunately the issue is still present. One of the compute pods logs: 

```
default computed: 2022-05-13T19:46:31.549390Z  INFO computed: coordinator connection accepted
default thread 'timely:recv-30' panicked at 'assertion failed: read > 0', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/7ad8962/communication/src/allocator/zero_copy/tcp.rs:67:9
default note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
default thread 'timely:work-0' panicked at 'MergeQueue poisoned.', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/7ad8962/communication/src/allocator/zero_copy/bytes_exchange.rs:96:48
default computed: 2022-05-13T19:46:31.552126Z  INFO computed: coordinator connection gone; lingering
default thread 'timely:send-0' panicked at 'MergeQueue poisoned.', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/7ad8962/communication/src/allocator/zero_copy/bytes_exchange.rs:96:48```

**Comment by benesch** [comment]
Fixed by #13388.
